integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/VisitorFactory.java

```diff
@@ -0,0 +1,23 @@
+package io.openlineage.spark.agent.lifecycle.plan;
+
+import io.openlineage.client.OpenLineage;
+import java.util.List;
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import scala.PartialFunction;
+
+/**
+ * Provides Visitors for iterating on {@link LogicalPlan}.
+ *
+ * <p>All common Visitors should be grouped and passed to {@link
+ * VisitorFactory#getOutputVisitors(List)} to retrieve Visitors for {@link
+ * OpenLineage.OutputDataset}
+ */
+public interface VisitorFactory {
+
+  List<PartialFunction<LogicalPlan, List<OpenLineage.Dataset>>> getCommonVisitors();
```

Comment:
Can we start using the `QueryPlanVisitor` type rather than referencing the `PartialFunction` type everywhere? I feel that that is clearer to readers.

Reply:
`QueryPlanVisitor<T extends LogicalPlan> extends AbstractPartialFunction<LogicalPlan, List<OpenLineage.Dataset>>`
In our case we need` List<OpenLineage.InputDataset> ` and `List<OpenLineage.OutputDataset>`

Reply:
We can consider changing the `QueryPlanVisitor` signature to
```java 
QueryPlanVisitor<T extends LogicalPlan, D extends OpenLineage.Dataset> extends AbstractPartialFunction<LogicalPlan, List<D>>
```

Reply:
I've changed it